Fix ChainerX CBLAS link issue

grafi-tt opened the pull request:
This should fix #6607.

@HiroakiMikami Could you try ChainerX build with this branch? I don't have Arch Linux environment.

Probably the cause of the error is Arch Linux's OpenBLAS is built [without CBLAS API](https://git.archlinux.org/svntogit/community.git/tree/trunk/PKGBUILD?h=packages/openblas#n22). CBLAS functions are separately provided in libcblas package.

This PR

* guards `<cblas.h>` by `extern "C"` and
* searches `libcblas.so` in `FindCBLAS.cmake` if required.

Those changes also fixes the issue with ATLAS on Ubuntu 18.04. ATLAS ships `cblas.h` without `__cplusplus` checking, so guards should be added at include place.

Note that the PR don't address header file detection. Notably MKL wouldn't work since it ships `mkl_cblas.h`. Though it is a different issue, I think it is better to bundle our own `cblas.h` in ChainerX as [numpy does](https://github.com/numpy/numpy/blob/master/numpy/core/src/common/npy_cblas.h).

HiroakiMikami commented on chainerx_cc/cmake/FindBLAS.cmake:
```diff
@@ -171,6 +171,21 @@ macro(Check_Fortran_Libraries LIBRARIES _prefix _name _flags _list _thread)
+    if(NOT ${_prefix}${_combined_name}_HAS_CBLAS)
+        find_library(${_prefix}${_combined_name}_CBLAS_LIBRARY cblas)
+        if(${_prefix}${_combined_name}_CBLAS_LIBRARY)
+          set(${LIBRARIES} ${${LIBRARIES}} ${_prefix}${_combined_name}_CBLAS_LIBRARY)
```
This line appends the variable name (e.g., BLAS_openblas_CBLAS_LIBRARY) to the `LIBRARIES` variable.
`set(${LIBRARIES} ${${LIBRARIES}} ${${_prefix}${_combined_name}_CBLAS_LIBRARY})` seems correct.

HiroakiMikami commented:
This PR almost works for me, but there is one trivial mistake I commented.

grafi-tt commented:
@HiroakiMikami Thank you for pointing out my silly mistake. I fixed the mistake and confirmed that `FindBLAS` works well.

@hvy Could you review the patch?

grafi-tt commented:
ping

hvy commented:
Sorry for the delayed review. I'll try to take a look at it today.

hvy commented:
Jenkins, test this please.

hvy commented on chainerx_cc/cmake/FindBLAS.cmake:
```diff
@@ -171,6 +171,21 @@ macro(Check_Fortran_Libraries LIBRARIES _prefix _name _flags _list _thread)
+        if(${_prefix}${_combined_name}_CBLAS_LIBRARY)
+          set(${LIBRARIES} ${${LIBRARIES}} ${${_prefix}${_combined_name}_CBLAS_LIBRARY})
+        else()
+          set(_libraries_work FALSE)
```
Could you explain why this is needed?

grafi-tt replied:
* If `cblas_sgemm` symbol not found in the blas library (e.g. `libopenblas.so`), then `${_prefix}${_combined_name}_HAS_CBLAS` is set false.
* In this case, check whether `libcblas.so` exists. This file should provides CBLAS API.
* If `libcblas.so` not found, then we couldn't use CBLAS API (only FORTRAN API available.)  So we revert `_li
brafy_work` flag.

This situation may happen when user is running Arch Linux. It strips CBLAS API from OpenBLAS, and provides the API by a separate package [`cblas`](https://www.archlinux.org/packages/extra/x86_64/cblas/) that contains `libcblas.so`.

hvy replied:
Thanks for the explanation. I just want to make sure but the vendor for which this path is passed is OpenBLAS? Is it the case that no other vendors are available (e.g. ATLAS) in the environment for which this issue tries to address? If not, you could use `BLA_VENDOR ` to specify the candidate vendors explicitly. For reference, this seems like a step towards a less ad-hoc approach to me, but this also depends on the answers to those questions. https://cmake-basis.github.io/apidoc/latest/FindBLAS_8cmake_source.html

grafi-tt replied:
ATLAS has the same issue. ATLAS officially provides `libatlas.so` and `libcblas.so` separately.

https://packages.ubuntu.com/ja/bionic/amd64/libatlas3-base/filelist

Debian and Ubuntu also build `libblas.so` to support BLAS library switching. Since we blacklisted `libatlas.so`, the issue hadn't appear on those distributions.

https://salsa.debian.org/science-team/atlas/blob/a53f04c697a75dd39487fe331d49eea13effefa7/debian/rules#L178-182

hvy replied:
I see. I just find the current change a bit ad-hoc since it's non-trivial and not visible outside `Check_Fortran_Libraries`. How about adding another argument to it, for additional libraries to search for (and changing the `_name` passed to the function for OpenBLAS and ATLAS to the "missing" symbol?

hvy commented:
Jenkins, test this please.

hvy commented:
In case the comment got slipped away, could you address https://github.com/chainer/chainer/pull/6639#discussion_r289241981 @grafi-tt ?

hvy commented:
@HiroakiMikami could you check if the current changes solves your issue?

hvy commented:
Jenkins, test this please.

grafi-tt commented:
Sorry for late response. I rebased the changes onto master, and squashed commits for simpler history.

The test failure is not related to the change. Could you please check again? @hvy @HiroakiMikami
